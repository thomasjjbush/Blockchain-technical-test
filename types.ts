import { AnyAction } from 'redux';
import { ThunkAction, ThunkDispatch } from 'redux-thunk';
import { Json } from 'enzyme-to-json';
import HtmlWebPackPlugin from 'html-webpack-plugin';

export type Action = ThunkAction<Promise<AnyAction> | AnyAction, Store, unknown, AnyAction>;

export interface Address {
    address?: string;
    error?: boolean;
    final_balance?: number;
    hash160?: string;
    n_tx?: number;
    total_received?: number;
    total_sent?: number;
    tx_per_page: number;
    txs?: Txs[];
}

export enum AddressActions {
    ERROR = 'address/ERROR',
    LOADED = 'address/LOADED',
}

export type Dispatch = ThunkDispatch<Store, void, AnyAction>;

export type FetchParams = Record<string, boolean | number | string>;

export interface Home {
    error?: boolean;
    latest_transactions: Txs[];
}

export enum HomeActions {
    ERROR = 'home/ERROR',
    LATEST_TRANSACTION = 'home/LATEST_TRANSACTION',
}

export interface MultiPageButtonsProps {
    currentPage: number;
    numberOfButtons: number;
    numberOfPages: number;
    skip?: number;
    to: (i: number) => Partial<Location>;
}

export interface SearchInputProps {
    onChange: (text: string) => void;
    placeholder?: string;
    value: string;
}

export interface SerializerMap extends Json {
    node: {
        type: {
            componentStyle?: {
                baseStyle?: {
                    rules: (string | ((props: StyledProps) => string))[];
                };
                rules: (string | ((props: StyledProps) => string))[];
            };
        };
    };
}

export interface SocketOptions<T> {
    onClose: string;
    onError: T;
    onMessage: T;
    onOpen: string;
}

export interface Store {
    address: Address;
    home: Home;
}

export type StyledProps<Props = {}> = { theme: Theme } & Props;

export interface Theme {
    colors: {
        brand: string;
        offset: string;
    };
}

export interface TransactionProps extends Partial<Txs> {
    search?: string;
}

export interface Txs {
    balance: number;
    fee: number;
    hash: string;
    result: number;
    size: number;
}

export interface WebpackConfig {
    devServer: { historyApiFallback: boolean; hot: boolean; open: boolean };
    entry: string[];
    mode: string;
    module: {
        rules: WebpackRule[];
    };
    output: { path?: string; publicPath: string };
    plugins: HtmlWebPackPlugin[];
    resolve: { extensions: string[] };
}

interface WebpackRule {
    exclude?: (RegExp | string)[];
    test: RegExp;
    use: {
        loader: string;
        options?: { plugins: string[] };
    };
}
