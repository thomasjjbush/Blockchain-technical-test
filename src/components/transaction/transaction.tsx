import React from 'react';
import styled from 'styled-components';
import { FC, ReactElement } from 'react';
import { highlightTerm } from '../../utils/highlight-term/highlight-term';
import { StyledProps, TransactionProps as Props } from '../../../types';

const Hash = styled.h3<StyledProps>`
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;

    span {
        background-color: ${({ theme }): string => theme.colors.brand};
        color: white;
    }
`;

const Container = styled.div<StyledProps>`
    border-bottom: solid ${(props): string => props.theme.colors.offset};
`;

export const Transaction: FC<Props> = ({ fee, hash, search, size }: Props): ReactElement => {
    return (
        <Container>
            <Hash data-test-id="hash">Hash: {highlightTerm(hash, search)}</Hash>
            {fee && <h3>Fee: {fee}</h3>}
            {size && <h3>Size: {size}</h3>}
        </Container>
    );
};
