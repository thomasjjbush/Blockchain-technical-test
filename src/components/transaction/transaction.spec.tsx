import React from 'react';
import { shallow } from 'enzyme';
import { Transaction } from './transaction';
import { TransactionProps as Props } from '../../../types';

describe('Transaction', () => {
    const props: Props = {
        hash: 'hash',
        fee: 20,
        size: 50,
    };

    it('should match snapshot', () => {
        expect(shallow(<Transaction {...props} />)).toMatchSnapshot();
    });

    it('should highlight term if it is provided and exists inside hash', () => {
        const wrapper = shallow(<Transaction {...props} search="ash" />);
        expect(wrapper.find({ 'data-test-id': 'hash' }).find('span').text()).toBe('ash');
    });
});
