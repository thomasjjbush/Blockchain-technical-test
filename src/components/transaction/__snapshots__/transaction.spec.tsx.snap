// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transaction should match snapshot 1`] = `
<styled.div
  styles="
    border-bottom: solid theme.colors.offset;
"
>
  <styled.h3
    data-test-id="hash"
    styles="
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;

    span {
        background-color: theme.colors.brand;
        color: white;
    }
"
  >
    Hash: 
    hash
  </styled.h3>
  <h3>
    Fee: 
    20
  </h3>
  <h3>
    Size: 
    50
  </h3>
</styled.div>
`;
